src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -189,12 +189,55 @@
         if ( authz.checkIfHasPermissions( new AclKey( entitySetId ),
                 Principals.getCurrentPrincipals(),
                 EnumSet.of( Permission.READ ) ) ) {
-            return loadNormalEntitySetData( entitySetId, selection );
+            EntitySet es = edmService.getEntitySet( entitySetId );
+            if(es.isLinking()) {
+                return loadLinkedEntitySetData( entitySetId, selection, es );
+            } else {
+                return loadNormalEntitySetData( entitySetId, selection );
+            }
         } else {
             throw new ForbiddenException( "Insufficient permissions to read the entity set or it doesn't exists." );
         }
     }
 
+    private EntitySetData<FullQualifiedName> loadLinkedEntitySetData(
+            UUID linkingEntitySetId,
+            EntitySetSelection selection,
+            EntitySet es) {
+        Set<UUID> allEntitySetIds = Sets.newHashSet( es.getLinkedEntitySets());
+        checkState( !allEntitySetIds.isEmpty(),
+                "Linked entity sets are empty for linking entity set %s", linkingEntitySetId );
+
+        EntityType linkedEntityType = edmService.getEntityTypeByEntitySetId( allEntitySetIds.iterator().next() );
+        checkState( allEntitySetIds.stream().skip( 1 )
+                        .map( esId -> edmService.getEntityTypeByEntitySetId( esId ) )
+                        .allMatch( linkedEntityType::equals ),
+                "Linked entity sets are of differing entity types {}", allEntitySetIds );
+
+        // In linked entity set calls, the selected entity key ids represent linking ids
+        Optional<Set<UUID>> linkingIds = selection.getEntityKeyIds();
+
+        return loadEntitySetData( getLinkedEntityKeyIds(allEntitySetIds, linkingIds), selection.getProperties());
+    }
+
+    private Map<UUID, Optional<Set<UUID>>> getLinkedEntityKeyIds(
+            Set<UUID> allEntitySetIds,
+            Optional<Set<UUID>> linkingIds) {
+
+        if(linkingIds.isPresent()) {
+            // if linkingIds is there: get all entityKey ids per entity set where linking_id =
+            Map<UUID, Set<UUID>> entityKeyIdsOfLinkingIds = linkingQueryService.getEntitiesWithLinkingIds(
+                    linkingIds.get());
+
+            return allEntitySetIds.stream().collect(Collectors.toMap(
+                    esId -> esId,
```

Comment:
Function.identity()

Reply:
Actually why not just ImmutableMap.copyOf(Maps.transformValues( entityKeyIdsOfLinkingIds, Optional::of) )?